inference-engine/src/multi_device/multi_device_exec_network.cpp

```diff
@@ -60,50 +90,150 @@ MultiDeviceExecutableNetwork::MultiDeviceExecutableNetwork(const DeviceMap<Infer
+        auto* workerRequestPtr = &workerRequest;
+        IE_ASSERT(idleWorkerRequests.try_push(workerRequestPtr) == true);
+        workerRequest._inferRequest->SetCallback(
+            [workerRequestPtr, this, device, idleWorkerRequestsPtr] (std::exception_ptr exceptionPtr) mutable {
+                IdleGuard idleGuard{workerRequestPtr, *idleWorkerRequestsPtr};
+                workerRequestPtr->_exceptionPtr = exceptionPtr;
+                {
+                    auto capturedTask = std::move(workerRequestPtr->_task);
+                    capturedTask();
+                }
+                // try to return the request to the idle list (fails if the overall object destruction has began)
+                if (idleGuard.Release()->try_push(workerRequestPtr)) {
+                    // let's try to pop a task, as we know there is at least one idle request, schedule if succeeded
+                    // if no device-agnostic tasks, let's try pop the device specific task, schedule if succeeded
+                    Task t;
+                    if (_inferPipelineTasks.try_pop(t))
+                        ScheduleToWorkerInferRequest(std::move(t));
+                    else if (_inferPipelineTasksDeviceSpecific[device]->try_pop(t))
+                        ScheduleToWorkerInferRequest(std::move(t), device);
+                }
+            });
+    }
+}
+
+MultiDeviceExecutableNetwork::MultiDeviceExecutableNetwork(const std::string&                         modelPath,
+                                                           const InferenceEngine::CNNNetwork&         network,
+                                                           const std::map<std::string, std::string>&  config,
+                                                           MultiDeviceInferencePlugin*                plugin)
+                                                           : _multiPlugin(plugin)
+                                                           , _workModeIsAUTO(true) {
+    if (_multiPlugin->GetCore() == nullptr) {
+        IE_THROW() << "Please, work with MULTI device via InferencEngine::Core object";
+    }
+
+    if (modelPath.empty() && network.getFunction() == nullptr) {
+        IE_THROW() << "MULTI device supports just ngraph network representation";
+    }
+
+    auto strDevices = _multiPlugin->GetDeviceList(config);
+    // collect the settings that are applicable to the devices we are loading the network to
+    _config[MultiDeviceConfigParams::KEY_MULTI_DEVICE_PRIORITIES] = strDevices;
+
+    auto metaDevices = _multiPlugin->ParseMetaDevices(strDevices, config);
+    _devicePrioritiesInitial = metaDevices;
+    _devicePriorities = metaDevices;
+
+    // init worker queue
+    for (auto&& device : metaDevices) {
+        _idleWorkerRequests[device.deviceName];
+    }
+
+    auto core = _multiPlugin->GetCore(); // shared_ptr that holds the Core while the lambda below (which captures that by val) works
+    auto LoadNetworkAsync =
```

Comment:
what confuses me is that the original async load to then devices in the MULTI happens in the plugin class:
https://github.com/openvinotoolkit/openvino/blob/master/inference-engine/src/multi_device/multi_device_plugin.cpp#L271   
For the 'AUTO' mode this happens in the ExecNetwork though. Shall it happen in one place? 
As we discussed when brainstormed the merging  - the only diff if the user-side LoadNetwork retruns upon first device loaded (AUTO), or all devices loaded (legacy MULTI beh). Shall we refactor these 2 items into one?

Reply:
Yes, I noticed this too, the reason that I move load action to ExecNetwork is that there are many members will be created in ExecNetwork, if I implement that in plugin class, there are will be too many parameters to construct ExecNetowork, see https://github.com/openvinotoolkit/openvino/pull/7076/files#diff-9d9984172782ce57b4ecae0570c917976cfcf12215927efafcee736688ce710aR154-R163 

I agreed to refactor this part, or I could move legacy loading action from plugin class to ExecNetwork class, what do you think?

Reply:
pls work with @apankratovantonp to review efficient async load to cover both cases.
e.g. MULTI's current load is good as it reuses the threads via the StreamExecutor (but it offers only wait for all), maybe we should user simple TBB code

Reply:
I use StreamExecutor to run load network now